test/framework/cluster_helpers.go

```diff
@@ -159,21 +160,37 @@ func DeleteCluster(ctx context.Context, input DeleteClusterInput) {
 
 // WaitForClusterDeletedInput is the input for WaitForClusterDeleted.
 type WaitForClusterDeletedInput struct {
-	Getter  Getter
+	Client  client.Client
 	Cluster *clusterv1.Cluster
+	// ArtifactFolder, if set, clusters will be dumped if deletion times out
+	ArtifactFolder string
 }
 
 // WaitForClusterDeleted waits until the cluster object has been deleted.
 func WaitForClusterDeleted(ctx context.Context, input WaitForClusterDeletedInput, intervals ...interface{}) {
 	Byf("Waiting for cluster %s to be deleted", klog.KObj(input.Cluster))
+	// Note: dumpArtifacts is passed in as a func so it gets only executed if and after the Eventually failed.
 	Eventually(func() bool {
 		cluster := &clusterv1.Cluster{}
 		key := client.ObjectKey{
 			Namespace: input.Cluster.GetNamespace(),
 			Name:      input.Cluster.GetName(),
 		}
-		return apierrors.IsNotFound(input.Getter.Get(ctx, key, cluster))
-	}, intervals...).Should(BeTrue())
+		return apierrors.IsNotFound(input.Client.Get(ctx, key, cluster))
+	}, intervals...).Should(BeTrue(), func() string { return dumpArtifacts(ctx, input.Client, input.Cluster, input.ArtifactFolder) })
+}
+
+func dumpArtifacts(ctx context.Context, client client.Client, cluster *clusterv1.Cluster, artifactFolder string) string {
+	if artifactFolder != "" {
+		// Dump all Cluster API related resources to artifacts.
+		DumpAllResources(ctx, DumpAllResourcesInput{
+			Lister:    client,
+			Namespace: cluster.Namespace,
+			LogPath:   filepath.Join(artifactFolder, "clusters-afterDeletionTimedOut", cluster.Name, "resources"),
```

Comment:
q: what if we are bumping more clusters, will they merge in the same folder?

Reply:
They have a different cluster.Name and because of that should get their own directory in the dump :-)